In [786]:
#Test commit
#Adding commit

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from datetime import datetime
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, RidgeCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import mean_squared_error, accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB


# Data Ingestion, Summary and Investigation

In [787]:
df = pd.read_csv('googleplaystore.csv')

In [788]:
df.head(10)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   
5                         Paper flowers instructions  ART_AND_DESIGN     4.4   
6            Smoke Effect Photo Maker - Smoke Editor  ART_AND_DESIGN     3.8   
7                                   Infinite Painter  ART_AND_DESIGN     4.1   
8                               Garden Coloring Book  ART_AND_DESIGN     4.4   
9                      Kids Paint Free - Drawing Fun  ART_AND_DESIGN     4.7   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   
5     167  5.6M      50,000+  Free     0       Everyone   
6     178   19M      50,000+  Free     0       Everyone   
7   36815   29M   1,000,000+  Free     0       Everyone   
8   13791   33M   1,000,000+  Free     0       Everyone   
9     121  3.1M      10,000+  Free     0       Everyone   

                      Genres        Last Updated         Current Ver  \
0               Art & Design     January 7, 2018               1.0.0   
1  Art & Design;Pretend Play    January 15, 2018               2.0.0   
2               Art & Design      August 1, 2018               1.2.4   
3               Art & Design        June 8, 2018  Varies with device   
4    Art & Design;Creativity       June 20, 2018                 1.1   
5               Art & Design      March 26, 2017                 1.0   
6               Art & Design      April 26, 2018                 1.1   
7               Art & Design       June 14, 2018            6.1.61.1   
8               Art & Design  September 20, 2017               2.9.2   
9    Art & Design;Creativity        July 3, 2018                 2.8   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up  
5    2.3 and up  
6  4.0.3 and up  
7    4.2 and up  
8    3.0 and up  
9  4.0.3 and up

In [789]:
df.tail()

App             Category  \
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10836       Everyone          Education     July 25, 2017                1.48   
10837       Everyone          Education      July 6, 2018                 1.0   
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

In [790]:
df.describe()

Rating
count  9367.000000
mean      4.193338
std       0.537431
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

In [791]:
df.shape

(10841, 13)

In [792]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [793]:
for column in df.columns:
    print(column.upper())
    print(df[column].unique())
    print('')
    print(df[column].value_counts())
    print('')

APP
['Photo Editor & Candy Camera & Grid & ScrapBook' 'Coloring book moana'
 'U Launcher Lite – FREE Live Cool Themes, Hide Apps' ...
 'Parkinson Exercices FR' 'The SCP Foundation DB fr nn5n'
 'iHoroscope - 2018 Daily Horoscope & Astrology']

ROBLOX                                                9
CBS Sports App - Scores, News, Stats & Watch Live     8
ESPN                                                  7
Duolingo: Learn Languages Free                        7
Candy Crush Saga                                      7
                                                     ..
Meet U - Get Friends for Snapchat, Kik & Instagram    1
U-Report                                              1
U of I Community Credit Union                         1
Waiting For U Launcher Theme                          1
iHoroscope - 2018 Daily Horoscope & Astrology         1
Name: App, Length: 9660, dtype: int64

CATEGORY
['ART_AND_DESIGN' 'AUTO_AND_VEHICLES' 'BEAUTY' 'BOOKS_AND_REFERENCE'
 'BUSINESS' 'COMICS' 'CO

# Data Cleaning and Imputation

In [794]:
def produce_missing_report(df):
    percent_missing= df.isnull().sum() *100 /len(df)
    missing_values_df=pd.DataFrame(
        {
        'column_name':df.columns,
        'percent_missing': percent_missing
        }
    )
    return missing_values_df

In [795]:
produce_missing_report(df)

column_name  percent_missing
App                        App         0.000000
Category              Category         0.000000
Rating                  Rating        13.596532
Reviews                Reviews         0.000000
Size                      Size         0.000000
Installs              Installs         0.000000
Type                      Type         0.009224
Price                    Price         0.000000
Content Rating  Content Rating         0.009224
Genres                  Genres         0.000000
Last Updated      Last Updated         0.000000
Current Ver        Current Ver         0.073794
Android Ver        Android Ver         0.027673

In [796]:
df[df['Rating']<1]

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

There are no rows with a rating less than one. Ratings must be between 1 and 5

In [797]:
invalid_rating= df[df['Rating']>5]
invalid_rating

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

There is one row with a rating more than 5 so we drop it.

In [798]:
df=df.drop(invalid_rating.index)

In [799]:
missing_ratings=df[df['Rating'].isnull()]
missing_ratings

App             Category  Rating  \
23                  Mcqueen Coloring pages       ART_AND_DESIGN     NaN   
113              Wrinkles and rejuvenation               BEAUTY     NaN   
123                 Manicure - nail design               BEAUTY     NaN   
126           Skin Care and Natural Beauty               BEAUTY     NaN   
129    Secrets of beauty, youth and health               BEAUTY     NaN   
...                                    ...                  ...     ...   
10824                            Cardio-FR              MEDICAL     NaN   
10825                   Naruto & Boruto FR               SOCIAL     NaN   
10831             payermonstationnement.fr  MAPS_AND_NAVIGATION     NaN   
10835                             FR Forms             BUSINESS     NaN   
10838               Parkinson Exercices FR              MEDICAL     NaN   

      Reviews  Size  Installs  Type Price Content Rating  \
23         61  7.0M  100,000+  Free     0       Everyone   
113       182  5.7M  100,000+  Free     0   Everyone 10+   
123       119  3.7M   50,000+  Free     0       Everyone   
126       654  7.4M  100,000+  Free     0           Teen   
129        77  2.9M   10,000+  Free     0     Mature 17+   
...       ...   ...       ...   ...   ...            ...   
10824      67   82M   10,000+  Free     0       Everyone   
10825       7  7.7M      100+  Free     0           Teen   
10831      38  9.8M    5,000+  Free     0       Everyone   
10835       0  9.6M       10+  Free     0       Everyone   
10838       3  9.5M    1,000+  Free     0       Everyone   

                                Genres        Last Updated Current Ver  \
23     Art & Design;Action & Adventure       March 7, 2018       1.0.0   
113                             Beauty  September 20, 2017         8.0   
123                             Beauty       July 23, 2018         1.3   
126                             Beauty       July 17, 2018        1.15   
129                             Beauty      August 8, 2017         2.0   
...                                ...                 ...         ...   
10824                          Medical       July 31, 2018       2.2.2   
10825                           Social    February 2, 2018         1.0   
10831                Maps & Navigation       June 13, 2018   2.0.148.0   
10835                         Business  September 29, 2016       1.1.5   
10838                          Medical    January 20, 2017         1.0   

      Android Ver  
23     4.1 and up  
113    3.0 and up  
123    4.1 and up  
126    4.1 and up  
129    2.3 and up  
...           ...  
10824  4.4 and up  
10825  4.0 and up  
10831  4.0 and up  
10835  4.0 and up  
10838  2.2 and up  

[1474 rows x 13 columns]

In [800]:
missing_ratings=df['Rating'].isnull()
none_missing= ~missing_ratings
df=df[none_missing]
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9366 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             9366 non-null   object 
 1   Category        9366 non-null   object 
 2   Rating          9366 non-null   float64
 3   Reviews         9366 non-null   object 
 4   Size            9366 non-null   object 
 5   Installs        9366 non-null   object 
 6   Type            9366 non-null   object 
 7   Price           9366 non-null   object 
 8   Content Rating  9366 non-null   object 
 9   Genres          9366 non-null   object 
 10  Last Updated    9366 non-null   object 
 11  Current Ver     9362 non-null   object 
 12  Android Ver     9364 non-null   object 
dtypes: float64(1), object(12)
memory usage: 1.0+ MB


In [801]:
produce_missing_report(df)

column_name  percent_missing
App                        App         0.000000
Category              Category         0.000000
Rating                  Rating         0.000000
Reviews                Reviews         0.000000
Size                      Size         0.000000
Installs              Installs         0.000000
Type                      Type         0.000000
Price                    Price         0.000000
Content Rating  Content Rating         0.000000
Genres                  Genres         0.000000
Last Updated      Last Updated         0.000000
Current Ver        Current Ver         0.042708
Android Ver        Android Ver         0.021354

The rows containing missing values for Type and Content Rating were also dropped

In [802]:
df['Current Ver'].value_counts()

Varies with device    1415
1.0                    458
1.1                    195
1.2                    126
1.3                    120
                      ... 
2.9.10                   1
3.18.5                   1
1.3.A.2.9                1
9.9.1.1910               1
0.3.4                    1
Name: Current Ver, Length: 2638, dtype: int64

Varies with device is the mode value. We will replace missing values with it.

In [803]:
df[df['Current Ver'].isnull()]

App            Category  Rating Reviews  \
15    Learn To Draw Kawaii Characters      ART_AND_DESIGN     3.2      55   
1553             Market Update Helper  LIBRARIES_AND_DEMO     4.1   20145   
6322           Virtual DJ Sound Mixer               TOOLS     4.2    4010   
7333                      Dots puzzle              FAMILY     4.0     179   

      Size    Installs  Type  Price Content Rating            Genres  \
15    2.7M      5,000+  Free      0       Everyone      Art & Design   
1553   11k  1,000,000+  Free      0       Everyone  Libraries & Demo   
6322  8.7M    500,000+  Free      0       Everyone             Tools   
7333   14M     50,000+  Paid  $0.99       Everyone            Puzzle   

           Last Updated Current Ver Android Ver  
15         June 6, 2018         NaN  4.2 and up  
1553  February 12, 2013         NaN  1.5 and up  
6322       May 10, 2017         NaN  4.0 and up  
7333     April 18, 2018         NaN  4.0 and up

In [804]:
df['Current Ver'].mode()

0    Varies with device
Name: Current Ver, dtype: object

In [805]:
df['Current Ver'] = df['Current Ver'].fillna('Varies with device')
df['Current Ver'].value_counts()

Varies with device    1419
1.0                    458
1.1                    195
1.2                    126
1.3                    120
                      ... 
2.9.10                   1
3.18.5                   1
1.3.A.2.9                1
9.9.1.1910               1
0.3.4                    1
Name: Current Ver, Length: 2638, dtype: int64

In [806]:
produce_missing_report(df)

column_name  percent_missing
App                        App         0.000000
Category              Category         0.000000
Rating                  Rating         0.000000
Reviews                Reviews         0.000000
Size                      Size         0.000000
Installs              Installs         0.000000
Type                      Type         0.000000
Price                    Price         0.000000
Content Rating  Content Rating         0.000000
Genres                  Genres         0.000000
Last Updated      Last Updated         0.000000
Current Ver        Current Ver         0.000000
Android Ver        Android Ver         0.021354

In [807]:
df['Android Ver'].value_counts()

4.1 and up            2059
Varies with device    1319
4.0.3 and up          1240
4.0 and up            1131
4.4 and up             875
2.3 and up             582
5.0 and up             535
4.2 and up             338
2.3.3 and up           240
3.0 and up             211
2.2 and up             208
4.3 and up             207
2.1 and up             113
1.6 and up              87
6.0 and up              48
7.0 and up              41
3.2 and up              31
2.0 and up              27
5.1 and up              18
1.5 and up              16
3.1 and up               8
2.0.1 and up             7
4.4W and up              6
8.0 and up               5
7.1 and up               3
4.0.3 - 7.1.1            2
5.0 - 8.0                2
1.0 and up               2
7.0 - 7.1.1              1
4.1 - 7.1.1              1
5.0 - 6.0                1
Name: Android Ver, dtype: int64

In [808]:
df[df['Android Ver'].isnull()]

App         Category  Rating Reviews  Size Installs  \
4453  [substratum] Vacuum: P  PERSONALIZATION     4.4     230   11M   1,000+   
4490    Pi Dark [substratum]  PERSONALIZATION     4.5     189  2.1M  10,000+   

      Type  Price Content Rating           Genres    Last Updated Current Ver  \
4453  Paid  $1.49       Everyone  Personalization   July 20, 2018         4.4   
4490  Free      0       Everyone  Personalization  March 27, 2018         1.1   

     Android Ver  
4453         NaN  
4490         NaN

Both rows containing missing values were last updated in 2018. We will replace the adroid ver with the modal adroid ver in 2018

In [809]:
df2018=df[df['Last Updated'].str.contains('2018')]
df2018['Android Ver'].mode()    #modal Adroid Ver of apps from 2018

0    4.1 and up
Name: Android Ver, dtype: object

In [810]:
df['Android Ver'] = df['Android Ver'].fillna('4.1 and up')
df['Android Ver'].value_counts()

4.1 and up            2061
Varies with device    1319
4.0.3 and up          1240
4.0 and up            1131
4.4 and up             875
2.3 and up             582
5.0 and up             535
4.2 and up             338
2.3.3 and up           240
3.0 and up             211
2.2 and up             208
4.3 and up             207
2.1 and up             113
1.6 and up              87
6.0 and up              48
7.0 and up              41
3.2 and up              31
2.0 and up              27
5.1 and up              18
1.5 and up              16
3.1 and up               8
2.0.1 and up             7
4.4W and up              6
8.0 and up               5
7.1 and up               3
4.0.3 - 7.1.1            2
5.0 - 8.0                2
1.0 and up               2
7.0 - 7.1.1              1
4.1 - 7.1.1              1
5.0 - 6.0                1
Name: Android Ver, dtype: int64

In [811]:
produce_missing_report(df)

column_name  percent_missing
App                        App              0.0
Category              Category              0.0
Rating                  Rating              0.0
Reviews                Reviews              0.0
Size                      Size              0.0
Installs              Installs              0.0
Type                      Type              0.0
Price                    Price              0.0
Content Rating  Content Rating              0.0
Genres                  Genres              0.0
Last Updated      Last Updated              0.0
Current Ver        Current Ver              0.0
Android Ver        Android Ver              0.0

In [812]:
df.drop_duplicates(keep = 'first', ignore_index = True)
df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10834     4.0       7                2.6M         500+  Free     0   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10834       Everyone                  Education     June 18, 2017   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10834               1.0.0          4.1 and up  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[9366 rows x 13 columns]

In [813]:
df[['Size', 'Price', 'Installs']] = df[['Size', 'Price', 'Installs']].astype(str)
df['Reviews']= df['Reviews'].astype('int64')


In [814]:
def convert(size):
    convert=size.str.removesuffix("k")
    convert=convert.astype('float64')
    convert=convert*0.001
    return convert

In [815]:
clean_size= df['Size']
clean_size= clean_size.replace('Varies with device', '0.0')
clean_size= clean_size.str.removesuffix('M')
clean_size=clean_size.str.removesuffix('k')
df['Size']=clean_size
df['Size']=df['Size'].astype('float64')

In [816]:
clean_price= df['Price']
clean_price= clean_price.replace(0, 0.0)
clean_price=clean_price.str.removeprefix('$')
clean_price.value_counts()
df['Price']=clean_price
df['Price']=df['Price'].astype('float64')

In [817]:
def clean_strings(string, st_letter, ls_letter, caracter_br, caracter_af):
    string = string[st_letter: -ls_letter]
    string = string.replace(caracter_br, caracter_af)
    string = int(string)
    return string

In [818]:
clean_installs=df['Installs']
clean_installs= clean_installs.apply(lambda x: clean_strings(x, 0, 1, ',', ''))
clean_installs.value_counts()


1000000       1577
10000000      1252
100000        1150
10000         1010
5000000        752
1000           713
500000         538
50000          467
5000           432
100000000      409
100            309
50000000       289
500            201
500000000       72
10              69
1000000000      58
50              56
5                9
1                3
Name: Installs, dtype: int64

In [819]:
clean_installs.describe()

count    9.366000e+03
mean     1.789744e+07
std      9.123822e+07
min      1.000000e+00
25%      1.000000e+04
50%      5.000000e+05
75%      5.000000e+06
max      1.000000e+09
Name: Installs, dtype: float64

In [820]:
def normalize_the_installs(value):
    if value < 10**4:   # first quartile
        return '0+' 
    elif 10**4 <= value < 10**5:    # second quartile
        return '10,000+'
    elif 10**5 <= value < 10**6:    # third quartile
        return '100,000+'
    elif 10**6 <= value <10**8:    # fourth quartile
        return '1,000,000+'
    elif 10**8 <= value <10**9:    # fourth quartile
        return '100,000,000'
    else:   #if >10**9  , outliers
        return '1,000,000,000+'

In [821]:
df['Installs']=clean_installs
df['Installs'] = df['Installs'].apply(lambda x: normalize_the_installs(x))
df['Installs']

0           10,000+
1          100,000+
2        1,000,000+
3        1,000,000+
4          100,000+
            ...    
10834            0+
10836            0+
10837            0+
10839            0+
10840    1,000,000+
Name: Installs, Length: 9366, dtype: object

In [822]:
df['Installs'].value_counts()

1,000,000+        3870
0+                1792
100,000+          1688
10,000+           1477
100,000,000        481
1,000,000,000+      58
Name: Installs, dtype: int64

In [823]:
def get_year(lupdate):
    if lupdate.find(',') != -1 :  #only perform operation on Titles that contain a year
        components= lupdate.split(',')    
        Year = components[-1].strip()  #get year and one bracket
        #Year= Year[:-1] #remove 2nd bracket
        Year= int(Year)
        return Year

In [824]:
df['Last Updated']=df['Last Updated'].apply(get_year)
df['Last Updated']

0        2018
1        2018
2        2018
3        2018
4        2018
         ... 
10834    2017
10836    2017
10837    2018
10839    2015
10840    2018
Name: Last Updated, Length: 9366, dtype: int64

In [825]:
df['Type'].value_counts()

Free    8719
Paid     647
Name: Type, dtype: int64

In [826]:
def clean_type(x):
    if x=='Free':
        return 1
    else:
        return 0

In [827]:
df['Type']=df['Type'].apply(clean_type)
df['Type']

0        1
1        1
2        1
3        1
4        1
        ..
10834    1
10836    1
10837    1
10839    1
10840    1
Name: Type, Length: 9366, dtype: int64

In [828]:
df['Content Rating'].value_counts()

Everyone           7420
Teen               1084
Mature 17+          461
Everyone 10+        397
Adults only 18+       3
Unrated               1
Name: Content Rating, dtype: int64

In [829]:
df['Content Rating']=df['Content Rating'].replace('Unrated', 'Everyone') #replace with mode

In [830]:
df['Content Rating'].value_counts()

Everyone           7421
Teen               1084
Mature 17+          461
Everyone 10+        397
Adults only 18+       3
Name: Content Rating, dtype: int64

In [831]:
newClass1 = pd.Categorical(df['Content Rating'], ordered=True)
newClass1 = newClass1.rename_categories(['1.0', '2.0','3.0','4.0','5.0'])
newClass1.describe() 

counts     freqs
categories                  
1.0              3  0.000320
2.0           7421  0.792334
3.0            397  0.042387
4.0            461  0.049221
5.0           1084  0.115738

In [832]:
df['Content Rating'] = newClass1

In [833]:
df['Category'].value_counts()

FAMILY                 1747
GAME                   1097
TOOLS                   734
PRODUCTIVITY            351
MEDICAL                 350
COMMUNICATION           328
FINANCE                 323
SPORTS                  319
PHOTOGRAPHY             317
PERSONALIZATION         314
LIFESTYLE               314
BUSINESS                303
HEALTH_AND_FITNESS      297
SOCIAL                  259
SHOPPING                238
NEWS_AND_MAGAZINES      233
TRAVEL_AND_LOCAL        226
DATING                  195
BOOKS_AND_REFERENCE     178
VIDEO_PLAYERS           160
EDUCATION               155
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     124
FOOD_AND_DRINK          109
HOUSE_AND_HOME           76
WEATHER                  75
AUTO_AND_VEHICLES        73
LIBRARIES_AND_DEMO       65
ART_AND_DESIGN           62
COMICS                   58
PARENTING                50
EVENTS                   45
BEAUTY                   42
Name: Category, dtype: int64

In [834]:
newClass2 = pd.Categorical(df['Category'], ordered=True)
newClass2 = newClass2.rename_categories(['1.0', '2.0','3.0','4.0','5.0','6.0','7.0','8.0','9.0','10.0','11.0', '12.0','13.0', '14.0', '15.0', '16.0', '17.0', '18.0', '19.0', '20.0','21.0','22.0','23.0','24.0','25.0', '26.0','27.0','28.0','29.0','30.0','31.0','32.0','33.0'])
newClass2.describe()    

counts     freqs
categories                  
1.0             62  0.006620
2.0             73  0.007794
3.0             42  0.004484
4.0            178  0.019005
5.0            303  0.032351
6.0             58  0.006193
7.0            328  0.035020
8.0            195  0.020820
9.0            155  0.016549
10.0           149  0.015909
11.0            45  0.004805
12.0          1747  0.186526
13.0           323  0.034486
14.0           109  0.011638
15.0          1097  0.117126
16.0           297  0.031710
17.0            76  0.008114
18.0            65  0.006940
19.0           314  0.033526
20.0           124  0.013239
21.0           350  0.037369
22.0           233  0.024877
23.0            50  0.005338
24.0           314  0.033526
25.0           317  0.033846
26.0           351  0.037476
27.0           238  0.025411
28.0           259  0.027653
29.0           319  0.034059
30.0           734  0.078369
31.0           226  0.024130
32.0           160  0.017083
33.0            75  0.008008

In [835]:
df['Category'] = newClass2

In [836]:
df.head(5)

App Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook      1.0     4.1   
1                                Coloring book moana      1.0     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...      1.0     4.7   
3                              Sketch - Draw & Paint      1.0     4.5   
4              Pixel Draw - Number Art Coloring Book      1.0     4.3   

   Reviews  Size    Installs  Type  Price Content Rating  \
0      159  19.0     10,000+     1    0.0            2.0   
1      967  14.0    100,000+     1    0.0            2.0   
2    87510   8.7  1,000,000+     1    0.0            2.0   
3   215644  25.0  1,000,000+     1    0.0            5.0   
4      967   2.8    100,000+     1    0.0            2.0   

                      Genres  Last Updated         Current Ver   Android Ver  
0               Art & Design          2018               1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play          2018               2.0.0  4.0.3 and up  
2               Art & Design          2018               1.2.4  4.0.3 and up  
3               Art & Design          2018  Varies with device    4.2 and up  
4    Art & Design;Creativity          2018                 1.1    4.4 and up

In [837]:
new_df= df['Genres'].str.get_dummies(';') 
new_df

Action  Action & Adventure  Adventure  Arcade  Art & Design  \
0           0                   0          0       0             1   
1           0                   0          0       0             1   
2           0                   0          0       0             1   
3           0                   0          0       0             1   
4           0                   0          0       0             1   
...       ...                 ...        ...     ...           ...   
10834       0                   0          0       0             0   
10836       0                   0          0       0             0   
10837       0                   0          0       0             0   
10839       0                   0          0       0             0   
10840       0                   0          0       0             0   

       Auto & Vehicles  Beauty  Board  Books & Reference  Brain Games  ...  \
0                    0       0      0                  0            0  ...   
1                    0       0      0                  0            0  ...   
2                    0       0      0                  0            0  ...   
3                    0       0      0                  0            0  ...   
4                    0       0      0                  0            0  ...   
...                ...     ...    ...                ...          ...  ...   
10834                0       0      0                  0            0  ...   
10836                0       0      0                  0            0  ...   
10837                0       0      0                  0            0  ...   
10839                0       0      0                  1            0  ...   
10840                0       0      0                  0            0  ...   

       Simulation  Social  Sports  Strategy  Tools  Travel & Local  Trivia  \
0               0       0       0         0      0               0       0   
1               0       0       0         0      0               0       0   
2               0       0       0         0      0               0       0   
3               0       0       0         0      0               0       0   
4               0       0       0         0      0               0       0   
...           ...     ...     ...       ...    ...             ...     ...   
10834           0       0       0         0      0               0       0   
10836           0       0       0         0      0               0       0   
10837           0       0       0         0      0               0       0   
10839           0       0       0         0      0               0       0   
10840           0       0       0         0      0               0       0   

       Video Players & Editors  Weather  Word  
0                            0        0     0  
1                            0        0     0  
2                            0        0     0  
3                            0        0     0  
4                            0        0     0  
...                        ...      ...   ...  
10834                        0        0     0  
10836                        0        0     0  
10837                        0        0     0  
10839                        0        0     0  
10840                        0        0     0  

[9366 rows x 53 columns]

In [838]:
new_df.columns

Index(['Action', 'Action & Adventure', 'Adventure', 'Arcade', 'Art & Design',
       'Auto & Vehicles', 'Beauty', 'Board', 'Books & Reference',
       'Brain Games', 'Business', 'Card', 'Casino', 'Casual', 'Comics',
       'Communication', 'Creativity', 'Dating', 'Education', 'Educational',
       'Entertainment', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo', 'Lifestyle',
       'Maps & Navigation', 'Medical', 'Music', 'Music & Audio',
       'Music & Video', 'News & Magazines', 'Parenting', 'Personalization',
       'Photography', 'Pretend Play', 'Productivity', 'Puzzle', 'Racing',
       'Role Playing', 'Shopping', 'Simulation', 'Social', 'Sports',
       'Strategy', 'Tools', 'Travel & Local', 'Trivia',
       'Video Players & Editors', 'Weather', 'Word'],
      dtype='object')

In [839]:
merged_df= df.join(new_df)
merged_df

App Category  Rating  \
0         Photo Editor & Candy Camera & Grid & ScrapBook      1.0     4.1   
1                                    Coloring book moana      1.0     3.9   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...      1.0     4.7   
3                                  Sketch - Draw & Paint      1.0     4.5   
4                  Pixel Draw - Number Art Coloring Book      1.0     4.3   
...                                                  ...      ...     ...   
10834                                      FR Calculator     12.0     4.0   
10836                                   Sya9a Maroc - FR     12.0     4.5   
10837                   Fr. Mike Schmitz Audio Teachings     12.0     5.0   
10839                      The SCP Foundation DB fr nn5n      4.0     4.5   
10840      iHoroscope - 2018 Daily Horoscope & Astrology     19.0     4.5   

       Reviews  Size    Installs  Type  Price Content Rating  \
0          159  19.0     10,000+     1    0.0            2.0   
1          967  14.0    100,000+     1    0.0            2.0   
2        87510   8.7  1,000,000+     1    0.0            2.0   
3       215644  25.0  1,000,000+     1    0.0            5.0   
4          967   2.8    100,000+     1    0.0            2.0   
...        ...   ...         ...   ...    ...            ...   
10834        7   2.6          0+     1    0.0            2.0   
10836       38  53.0          0+     1    0.0            2.0   
10837        4   3.6          0+     1    0.0            2.0   
10839      114   0.0          0+     1    0.0            4.0   
10840   398307  19.0  1,000,000+     1    0.0            2.0   

                          Genres  ...  Simulation Social Sports  Strategy  \
0                   Art & Design  ...           0      0      0         0   
1      Art & Design;Pretend Play  ...           0      0      0         0   
2                   Art & Design  ...           0      0      0         0   
3                   Art & Design  ...           0      0      0         0   
4        Art & Design;Creativity  ...           0      0      0         0   
...                          ...  ...         ...    ...    ...       ...   
10834                  Education  ...           0      0      0         0   
10836                  Education  ...           0      0      0         0   
10837                  Education  ...           0      0      0         0   
10839          Books & Reference  ...           0      0      0         0   
10840                  Lifestyle  ...           0      0      0         0   

       Tools  Travel & Local  Trivia  Video Players & Editors  Weather  Word  
0          0               0       0                        0        0     0  
1          0               0       0                        0        0     0  
2          0               0       0                        0        0     0  
3          0               0       0                        0        0     0  
4          0               0       0                        0        0     0  
...      ...             ...     ...                      ...      ...   ...  
10834      0               0       0                        0        0     0  
10836      0               0       0                        0        0     0  
10837      0               0       0                        0        0     0  
10839      0               0       0                        0        0     0  
10840      0               0       0                        0        0     0  

[9366 rows x 66 columns]

# Preprocessing

### With merged_df

In [840]:
df1=merged_df
df1=df1.drop(columns=['Genres', 'App', 'Installs','Current Ver', 'Android Ver'])
df1

Category  Rating  Reviews  Size  Type  Price Content Rating  \
0          1.0     4.1      159  19.0     1    0.0            2.0   
1          1.0     3.9      967  14.0     1    0.0            2.0   
2          1.0     4.7    87510   8.7     1    0.0            2.0   
3          1.0     4.5   215644  25.0     1    0.0            5.0   
4          1.0     4.3      967   2.8     1    0.0            2.0   
...        ...     ...      ...   ...   ...    ...            ...   
10834     12.0     4.0        7   2.6     1    0.0            2.0   
10836     12.0     4.5       38  53.0     1    0.0            2.0   
10837     12.0     5.0        4   3.6     1    0.0            2.0   
10839      4.0     4.5      114   0.0     1    0.0            4.0   
10840     19.0     4.5   398307  19.0     1    0.0            2.0   

       Last Updated  Action  Action & Adventure  ...  Simulation  Social  \
0              2018       0                   0  ...           0       0   
1              2018       0                   0  ...           0       0   
2              2018       0                   0  ...           0       0   
3              2018       0                   0  ...           0       0   
4              2018       0                   0  ...           0       0   
...             ...     ...                 ...  ...         ...     ...   
10834          2017       0                   0  ...           0       0   
10836          2017       0                   0  ...           0       0   
10837          2018       0                   0  ...           0       0   
10839          2015       0                   0  ...           0       0   
10840          2018       0                   0  ...           0       0   

       Sports  Strategy  Tools  Travel & Local  Trivia  \
0           0         0      0               0       0   
1           0         0      0               0       0   
2           0         0      0               0       0   
3           0         0      0               0       0   
4           0         0      0               0       0   
...       ...       ...    ...             ...     ...   
10834       0         0      0               0       0   
10836       0         0      0               0       0   
10837       0         0      0               0       0   
10839       0         0      0               0       0   
10840       0         0      0               0       0   

       Video Players & Editors  Weather  Word  
0                            0        0     0  
1                            0        0     0  
2                            0        0     0  
3                            0        0     0  
4                            0        0     0  
...                        ...      ...   ...  
10834                        0        0     0  
10836                        0        0     0  
10837                        0        0     0  
10839                        0        0     0  
10840                        0        0     0  

[9366 rows x 61 columns]

In [841]:
df1.shape

(9366, 61)

In [842]:
scalar=StandardScaler()
scalar.fit(df1)

StandardScaler()

In [843]:
std_data=scalar.transform(df1)

In [844]:
features=std_data.T
features

array([[-2.02976786, -2.02976786, -2.02976786, ..., -0.70072954,
        -1.66730286,  0.14502212],
       [-0.17810357, -0.56630885,  0.98651225, ...,  1.56882016,
         0.59830697,  0.59830697],
       [-0.16345782, -0.16320081, -0.13567331, ..., -0.16350712,
        -0.16347213, -0.03681535],
       ...,
       [-0.13308499, -0.13308499, -0.13308499, ..., -0.13308499,
        -0.13308499, -0.13308499],
       [-0.08984614, -0.08984614, -0.08984614, ..., -0.08984614,
        -0.08984614, -0.08984614],
       [-0.05475857, -0.05475857, -0.05475857, ..., -0.05475857,
        -0.05475857, -0.05475857]])

In [845]:
covariance_matrix= np.cov(features)
covariance_matrix

array([[ 1.00010678e+00, -3.46070764e-02,  8.83121463e-04, ...,
         2.22439324e-01,  1.65022332e-01, -1.85248598e-02],
       [-3.46070764e-02,  1.00010678e+00,  6.81487432e-02, ...,
        -3.32152256e-02,  9.11175053e-03,  2.32749338e-02],
       [ 8.83121463e-04,  6.81487432e-02,  1.00010678e+00, ...,
         6.98886032e-03, -9.12656809e-03, -4.70625902e-03],
       ...,
       [ 2.22439324e-01, -3.32152256e-02,  6.98886032e-03, ...,
         1.00010678e+00, -1.19584492e-02, -7.28832187e-03],
       [ 1.65022332e-01,  9.11175053e-03, -9.12656809e-03, ...,
        -1.19584492e-02,  1.00010678e+00, -4.92037105e-03],
       [-1.85248598e-02,  2.32749338e-02, -4.70625902e-03, ...,
        -7.28832187e-03, -4.92037105e-03,  1.00010678e+00]])

In [846]:
eig_vals, eig_vecs= np.linalg.eig(covariance_matrix)
eig_vecs

array([[-6.70270989e-01,  6.36764467e-01, -3.26916850e-01, ...,
        -5.66123556e-04, -3.94160390e-03,  4.67816103e-04],
       [ 6.45396787e-02, -3.13327230e-04, -1.63129346e-04, ...,
         2.74938250e-03, -1.57664576e-02,  2.66606286e-03],
       [ 2.53136463e-02,  1.23748891e-03, -4.38790264e-04, ...,
        -5.33181851e-03,  6.16816736e-03,  8.81988467e-03],
       ...,
       [-1.40110617e-01, -7.85717392e-02,  1.90560228e-01, ...,
        -7.48839399e-03,  1.25148127e-01,  1.48668485e-01],
       [-1.08650701e-01, -6.25250322e-02,  1.34740125e-01, ...,
         3.50859800e-01, -8.16054775e-01,  2.56368365e-01],
       [ 9.40529552e-03,  3.79178804e-02,  4.35692875e-02, ...,
         3.65577612e-02, -9.50593329e-02, -5.07070075e-02]])

In [847]:
_sum= sum (eig_vals)
for _v in eig_vals:
    print(_v/_sum)

0.034484675386559464
0.00014444995458745452
5.085154358721644e-05
0.027451884041260022
0.007080881155154912
0.024462380552729066
0.009785462302435375
0.010314492866811064
0.022603127092519392
0.011237481609070859
0.021340977314686826
0.021224528117381757
0.012057939292264123
0.012441409178377615
0.02050912795234255
0.020198782874233913
0.012800405491517966
0.013258486508710785
0.01333261668152776
0.01962576989392464
0.019224060048566912
0.01861997784073225
0.014505845049847721
0.014813819759020002
0.018086617483650074
0.015116885026398524
0.017801943485717422
0.015448600905046277
0.017429750271777074
0.017239528346845383
0.01703357760357941
0.01701001227283874
0.016977046386330862
0.016970984711440745
0.01696163036614489
0.016925430647094678
0.016890568298766978
0.01687059028128645
0.01683071335979083
0.016798366056423222
0.01678620574503379
0.016779851750906825
0.016741080358985222
0.016705385676792524
0.01666466611262331
0.01664595071342459
0.016620310988038647
0.01642934409975778
0.

### With df

In [848]:
df2=df
df2=df2.drop(columns=['Genres', 'App', 'Installs','Current Ver', 'Android Ver'])
df2

Category  Rating  Reviews  Size  Type  Price Content Rating  \
0          1.0     4.1      159  19.0     1    0.0            2.0   
1          1.0     3.9      967  14.0     1    0.0            2.0   
2          1.0     4.7    87510   8.7     1    0.0            2.0   
3          1.0     4.5   215644  25.0     1    0.0            5.0   
4          1.0     4.3      967   2.8     1    0.0            2.0   
...        ...     ...      ...   ...   ...    ...            ...   
10834     12.0     4.0        7   2.6     1    0.0            2.0   
10836     12.0     4.5       38  53.0     1    0.0            2.0   
10837     12.0     5.0        4   3.6     1    0.0            2.0   
10839      4.0     4.5      114   0.0     1    0.0            4.0   
10840     19.0     4.5   398307  19.0     1    0.0            2.0   

       Last Updated  
0              2018  
1              2018  
2              2018  
3              2018  
4              2018  
...             ...  
10834          2017  
10836          2017  
10837          2018  
10839          2015  
10840          2018  

[9366 rows x 8 columns]

In [849]:
scalar=StandardScaler()
scalar.fit(df2)

StandardScaler()

In [850]:
std_data=scalar.transform(df2)

In [851]:
features=std_data.T
features

array([[-2.02976786, -2.02976786, -2.02976786, ..., -0.70072954,
        -1.66730286,  0.14502212],
       [-0.17810357, -0.56630885,  0.98651225, ...,  1.56882016,
         0.59830697,  0.59830697],
       [-0.16345782, -0.16320081, -0.13567331, ..., -0.16350712,
        -0.16347213, -0.03681535],
       ...,
       [-0.06075769, -0.06075769, -0.06075769, ..., -0.06075769,
        -0.06075769, -0.06075769],
       [-0.47748016, -0.47748016, -0.47748016, ..., -0.47748016,
         1.480523  , -0.47748016],
       [ 0.52607726,  0.52607726,  0.52607726, ...,  0.52607726,
        -2.19465142,  0.52607726]])

In [852]:
covariance_matrix= np.cov(features)
covariance_matrix

array([[ 1.00010678e+00, -3.46070764e-02,  8.83121463e-04,
        -1.96375565e-03, -2.32292593e-02, -1.29147751e-02,
        -1.06589196e-01, -2.42692702e-02],
       [-3.46070764e-02,  1.00010678e+00,  6.81487432e-02,
        -3.02668203e-02, -3.95852362e-02, -2.19056396e-02,
         1.99567905e-02,  1.37728727e-01],
       [ 8.83121463e-04,  6.81487432e-02,  1.00010678e+00,
        -5.51032829e-03,  4.32949528e-02, -9.82095736e-03,
         7.06471221e-02,  7.82930297e-02],
       [-1.96375565e-03, -3.02668203e-02, -5.51032829e-03,
         1.00010678e+00, -5.33541368e-02,  2.19953593e-02,
        -1.32283555e-02, -2.23516584e-01],
       [-2.32292593e-02, -3.95852362e-02,  4.32949528e-02,
        -5.33541368e-02,  1.00010678e+00, -2.23063657e-01,
         4.68116296e-02,  1.72490667e-01],
       [-1.29147751e-02, -2.19056396e-02, -9.82095736e-03,
         2.19953593e-02, -2.23063657e-01,  1.00010678e+00,
        -1.40725112e-02, -3.37911352e-03],
       [-1.06589196e-01,  1.995679

In [853]:
eig_vals, eig_vecs= np.linalg.eig(covariance_matrix)
eig_vecs

array([[-0.12743247,  0.0482029 , -0.18016642, -0.53077426,  0.21658525,
         0.56706776, -0.33553946,  0.43150254],
       [ 0.24659278,  0.30512891, -0.05576394, -0.20914653, -0.42088564,
         0.05014305,  0.6651999 ,  0.42081647],
       [ 0.2451384 ,  0.00703177,  0.27180834,  0.37682887, -0.19925377,
        -0.20298199, -0.47318333,  0.64743473],
       [-0.40404911, -0.30329361, -0.5842583 ,  0.07205523,  0.16393201,
        -0.44590909,  0.15655239,  0.38469509],
       [ 0.47456685,  0.5465195 , -0.41304689,  0.19087525,  0.50685908,
        -0.09098915, -0.05217291, -0.03397749],
       [-0.29692826,  0.39402532, -0.44917692,  0.18143755, -0.60643436,
         0.1160063 , -0.3211049 , -0.19382908],
       [ 0.23960926,  0.03011187, -0.03809146, -0.67217989, -0.18061549,
        -0.58928036, -0.29746967, -0.1415712 ],
       [ 0.5734211 , -0.59807201, -0.41877716,  0.078522  , -0.22820959,
         0.25864143, -0.04682589, -0.10387697]])

In [854]:
_sum= sum (eig_vals)
for _v in eig_vals:
    print(_v/_sum)

0.17812958775330434
0.08238098271251883
0.10215096125342897
0.10842373285995482
0.145149195658649
0.13736721686439987
0.11961503640959133
0.12678328648815274


Category: 17.8%
Rating: 8.2%
Reviews: 10.2%
Size: 10.8%
Type: 14.5%
Price: 13.7%
Content Rating: 11.9%
Last Updated: 12.6%

Category, Type and Price are the most important variables

In [855]:
df3=merged_df
x_data = df3.drop(['Genres', 'App', 'Installs','Current Ver', 'Android Ver','Rating', 'Reviews', 'Size', 'Content Rating', 'Last Updated'], axis = 1)
y_data = df3['Installs']

In [856]:
scaler_ = StandardScaler()
x_data = scaler_.fit_transform(x_data)

x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size = 0.2, random_state = 20)

# Models